parsely/src/main/java/com/parsely/parselyandroid/ParselyTracker.java

```diff
@@ -349,12 +351,16 @@ public void trackPlay(
                 videoEngagementManager = null;
             }
         }
+        @NonNull final String uuid = generatePixelId();
```

Comment:
Minor (🔍): Although adding `@NonNull` on a local variable doesn't hurt, it is kind of a redundancy as Java is at least smart enough to notice that a local variable maybe produce (or not) a NPE even without such a `@NonNull` or a `@Nullable` annotation. This is especially true since the `generatePixelId()` method is already annotated as `@NonNull`. Same applies to `buildEvent(...)` and the fact that its return type is also annotated as `@NonNull`.

FYI: From my experience these nullability annotation are mostly only used at constructor/method parameter and field level.

PS: Also, using final and a `@NonNull` is a further redundancy as this local variable can't be updated to `null` any longer anyway.

Reply:
Both `@NonNull` and `@Nullable` annotations have target of `LOCAL_VARIABLE` so I think it's fine using them here. You're right about `uuid` not being nullable is known from `generatePixelId` annotation, but at the same time, I'd have to look at `generatePixelId` signature to know it. Those annotations are not only for IDE, but also for people reading the code. I'd leave this as it is, as it doesn't hurt and helps in code readability IMO.

We can look at this from a different point: Kotlin equivalent of
```java
        @NonNull final String uuid = generatePixelId();
```
is
```
    val uuid: String = generatePixelId();
```

If we removed annotations and `final`, it'd be
```
    var uuid: String? = generatePixelId();
```

Reply:
Thanks for the reply @wzieba , as I mentioned this is minor and I am just glad it is being discussed! 🥇 

> ...only that final in this context is redundant!

Ah, actually I was saying that `@NonNull` is redundant, not the `final` keyword, I don't mind that, and to be honest, I kind of like it, doesn't hurt at all.

PS: Another Kotlin equivalent could have been one that you don't get a type reference at all: `val uuid = generatePixelId()`